File: app-runtime-management/agents/windows/src/TimeOnChrome.AppRuntime.Core/UsageSegment.cs
namespace TimeOnChrome.AppRuntime.Core;

public enum SegmentEndReason
{
    ApplicationSwitch,
    UserIdle,
    SessionUnavailable,
    SystemSleep,
    PeriodicSnapshot,
    StateCorrection,
    PipEnded,
    MediaStopped,
    CheckpointUnconfirmed,
    ServiceRecovery,
    ClockAdjustment,
    LateFact,
    Diagnostic,
}

public sealed record UsageSegment(
    string Id,
    string RuntimeSessionID,
    ApplicationIdentity Application,
    long StartAtMs,
    long EndAtMs,
    long DurationMilliseconds,
    SegmentEndReason EndReason);

File: macos-app-management/Sources/MacOSAppRuntimeCore/UsageSegment.swift
public enum SegmentEndReason: String, Codable, Equatable, Sendable {
    case applicationSwitch
    case userIdle
    case sessionUnavailable
    case systemSleep
    case periodicSnapshot
    case stateCorrection
}

public struct UsageSegment: Codable, Equatable, Sendable {
    public let id: String
    public let runtimeSessionID: String
    public let application: ApplicationIdentity
    public let startAtMs: Int64
    public let endAtMs: Int64
    public let durationMilliseconds: Int64
    public let endReason: SegmentEndReason

    public init(
        id: String,
        runtimeSessionID: String,
        application: ApplicationIdentity,
        startAtMs: Int64,
        endAtMs: Int64,
        durationMilliseconds: Int64,
        endReason: SegmentEndReason
    ) {
        self.id = id
        self.runtimeSessionID = runtimeSessionID
        self.application = application
        self.startAtMs = startAtMs
        self.endAtMs = endAtMs
        self.durationMilliseconds = durationMilliseconds
        self.endReason = endReason
    }
}
